src/methods/environment.js

```diff
@@ -459,33 +484,412 @@ export function ResolveThisBinding(realm: Realm): NullValue | ObjectValue | Abst
   return envRec.GetThisBinding();
 }
 
-export function BindingInitialization(realm: Realm, node: BabelNode, value: Value, environment: void | LexicalEnvironment) {
+export function BindingInitialization(realm: Realm, node: BabelNodeLVal, value: Value, strictCode: boolean, environment: void | LexicalEnvironment) {
   if (node.type === "ArrayPattern") { // ECMA262 13.3.3.5
     // 1. Let iterator be ? GetIterator(value).
+    let iterator = GetIterator(realm, value);
+
     // 2. Let iteratorRecord be Record {[[Iterator]]: iterator, [[Done]]: false}.
+    let iteratorRecord = {
+      $Iterator: iterator,
+      $Done: false,
+    };
+
     // 3. Let result be IteratorBindingInitialization for ArrayBindingPattern using iteratorRecord and environment as arguments.
+    try {
+      IteratorBindingInitialization(realm, node.elements, iteratorRecord, strictCode, environment);
+    } catch (error) {
+      // 4. If iteratorRecord.[[Done]] is false, return ? IteratorClose(iterator, result).
+      if (iteratorRecord.$Done === false && error instanceof AbruptCompletion) {
+        throw IteratorClose(realm, iterator, error);
+      }
+      throw error;
+    }
+
     // 4. If iteratorRecord.[[Done]] is false, return ? IteratorClose(iterator, result).
+    // TODO: Why does this work?
+    if (iteratorRecord.$Done === false) {
+      let completion = new AbruptCompletion(new UndefinedValue(realm));
+      let actualCompletion = IteratorClose(realm, iterator, completion);
+      if (actualCompletion !== completion) {
+        throw actualCompletion;
+      }
+    }
+
     // 5. Return result.
-    throw new Error("TODO: Patterns aren't supported yet");
+    return;
   } else if (node.type === "ObjectPattern") { // ECMA262 13.3.3.5
+    // BindingPattern : ObjectBindingPattern
+
     // 1. Perform ? RequireObjectCoercible(value).
     RequireObjectCoercible(realm, value);
 
     // 2. Return the result of performing BindingInitialization for ObjectBindingPattern using value and environment as arguments.
-    throw new Error("TODO: Patterns aren't supported yet");
+
+    for (let property of node.properties) {
```

Comment:
When the spec is obviously too complicated and there is a much simpler, obviously correct alternative, then use the second by all means. But in general, stick to the spec when it is at all reasonable to so so. Never deviate from the spec solely out of performance considerations.